File: src/RobotBounded.java
public class RobotBounded {
	public boolean isRobotBounded(String instructions) {

		if (instructions.length() == 0 || instructions == null)
			return false;

		int[][] dirs = { { 0, 1 }, { -1, 0 }, { 0, -1 }, { 1, 0 } };

		int n = 0; // north
		int x = 0;
		int y = 0; // initial robot co-ordinates
		for (int i = 0; i < instructions.length(); i++) {

			if ('L' == instructions.charAt(i)) {
				n = (n + 1) % 4;

			} else if ('R' == instructions.charAt(i)) {
				n = (n + 3) % 4;

			} else {
				x += dirs[n][0];
				y += dirs[n][1];
			}

		}
		if (n != 0 || (x == 0 && y == 0))
			return true;

		return false;

	}
}

File: RobotBounded.cpp
/*
Intuition: If the robot returns to its original position or if it is on it its original direction,
then the robot moves in a circle.

Simulate the robot position using a dirs array
Time Complexity: O(N), N = length of the instruction
Space Complexity: O(1)

*/
class Solution {
public:
    bool isRobotBounded(string instructions) {
        int currDirection = 0;
        int x = 0; int y = 0;
        vector<vector<int>> dirs = {{0,1}, {1,0}, {0,-1}, {-1,0}};
        for ( auto command: instructions){
            if ( command == 'L'){
                currDirection = (currDirection + 3) % 4;
            }
            else if ( command == 'R'){
                currDirection = (currDirection + 1) % 4;
            }
            else{
                x += dirs[currDirection][0];
                y += dirs[currDirection][1];
            }
        }
        return ( x== 0 and y == 0 ) or currDirection != 0;
    }
};